File: src/dynamic-programming/knight-probability-in-chessboard.cpp
class Solution {
public:
    vector<vector<int>> dirs = {{-2, -1}, {-2, 1}, {2, -1}, {2, 1}, {-1, -2}, {-1, 2}, {1, -2}, {1, 2}};
    double knightProbability(int n, int k, int row, int column) {
        //dp[k][row][column]表示从棋盘上任意一点出发后经过k次 能到达row column的概率之和 所以最后直接返回dp[k][row][column]
        vector<vector<vector<double>>> dp(k + 1, vector<vector<double>>(n,vector<double>(n)));
        
        for(int step = 0; step <= k; ++step) {
            for(int i = 0; i < n; ++i) {
                for(int j = 0; j < n; ++j) {
                    if(step == 0) {
                        dp[0][i][j] = 1.0;
                    }
                    else {
                        for(auto &dir : dirs) {
                            int ni = i + dir[0];
                            int nj = j + dir[1];

                            if(ni >= 0 && ni < n && nj >= 0 && nj <n) {
                                dp[step][i][j] += dp[step - 1][ni][nj] / 8;
                            }
                        }
                    }
                }
            }
        }

        return dp[k][row][column];
    }
};

File: src/dynamic-programming/knight-probability-in-chessboard.rs
const DIRS: [[i32; 2]; 8] = [
    [-2, -1],
    [-2, 1],
    [2, -1],
    [2, 1],
    [-1, -2],
    [-1, 2],
    [1, -2],
    [1, 2],
];

impl Solution {
    pub fn knight_probability(n: i32, k: i32, row: i32, column: i32) -> f64 {
        let n = n as usize;
        let k = k as usize;
        // dp[step][i][j]表示在走了step步且当前位置为(i, j)时还在棋盘内的概率
        let mut dp = vec![vec![vec![0.0; n]; n]; k + 1];

        for step in 0..=k {
            for i in 0..n {
                for j in 0..n {
                    if step == 0 {
                        dp[step][i][j] = 1.0;
                    } else {
                        for dir in &DIRS {
                            let ni = i as i32 + dir[0];
                            let nj = j as i32 + dir[1];
                            if ni >= 0 && ni < n as i32 && nj >= 0 && nj < n as i32 {
                                dp[step][i][j] += dp[step - 1][ni as usize][nj as usize] / 8.0;
                            }
                        }
                    }
                }
            }
        }

        dp[k][row as usize][column as usize]
    }
}
